content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
++++
+title = "Xi++: Object Oriented Features for Xi"
+[[extra.authors]]
+name = "Orko Sinha"
+link = ""
+[[extra.authors]]
+name = "Michael Maitland"
+link = "https://michaelmaitland.com"
++++
+Building a compiler for Xi is the main project in [CS 4120](https://www.cs.cornell.edu/courses/cs4120/) at Cornell. The final assignment, now discontinued, is to extend the compiler to support feaures of the [Xi++ Language Specification](https://www.cs.cornell.edu/courses/cs4120/2019sp/project/oolang.pdf) such as object-oriented programing. Our project is an implementation of these features on Michael's compiler created for CS 4120. Our project is [here](https://github.com/Yasgur99/xic/tree/oo).
+
+# Lexing
+For lexing, there are 5 new syntax elements that we need to handle, listed below:
+1. `class`
+2. `extends`
+3. `this`
+4. `new`
+5. `.` operator
+
+These correspond to 5 new token types corresponding to each keyword in the language. The existing JFlex file created for the original project is extended to support these features.
+
+# Parsing
+The parser handles the new semantics and allow parsing of the following:
+1. Parsing of `Interface` that contain zero or more `use` statements and zero or more class declarations. A class declaration is the name of the class, optionally the class name it extends, and zero or more function declarations. `ClassDecl` is created as a new AST node to represent this.
```

Comment:
TBH, I don't think the names of classes in your AST are all that relevant. Your reader will not be following along with your Java code, so names like `ClassDecl` are at the wrong level of abstraction for the reader to care about them.